### PR TITLE
ApplicationChoiceSubmission Validation needs attention

### PR DESCRIPTION
**Pull request by tomas-stefano**
## Context

An area of the code / application has thrown up bugs on two occassions and it should motivate us to review that area of the code closely.

#### What happened?

In October we updated the error messages (validation) on a candidates submission to be very fine grained based on specific attributes of their application to give them specific error messages and instructions.

This change resulted in a subtle bug that was only fixed in January.

After that fix, another issue arose.

This experience should tell us that this area of the code is not reliable and that we should decicate more resource to shore it up.

More system tests and find a way to make the logic clear and understandable, otherwise we are likely to introduce more bugs in future.

## Changes proposed in this pull request

* Grouping of all application choices in one spec
* Add more coverage (unit and feature specs)

## System specs coverage on this PR

For each validator:

|  Validator | System spec |
| ----------- | ----------- |
| immigration status | spec/system/candidate_interface/continuous_applications/submitting/candidate_with_no_right_to_work_tries_to_submit_course_spec.rb |
| applications closed   | spec/system/candidate_interface/continuous_applications/submitting/candidate_tries_to_submit_an_application_for_a_course_with_closed_application_spec.rb |
| course unavailable    | spec/system/candidate_interface/continuous_applications/submitting/candidate_tries_to_submit_an_application_for_course_unavailable_spec.rb |
| Primary/Science GCSE | spec/system/candidate_interface/continuous_applications/submitting/candidate_submitting_application_spec.rb |
| Incomplete details | spec/system/candidate_interface/continuous_applications/submitting/candidate_can_not_submit_with_incomplete_details_spec.rb | 

## Guidance to review

1. Are the specs enough? Do we need to increase coverage? If so, where?

## Link to Trello card

https://trello.com/c/IjD9RKWo/1208-bug-applicationchoicesubmission-validation-needs-attention

### CODE REVIEW

File: spec/support/test_helpers/candidate_submission_helper.rb [hunk 1]
```diff
@@ -0,0 +1,46 @@
+module CandidateSubmissionHelper
```
Comment (inulty-dfe) [inline spec/support/test_helpers/candidate_submission_helper.rb:1]: I think we need to continue discussing how we improve these helper modules. We need to be clear about what goes in them and what doesn't. I'm not sure adding this here now is helping us with that.

File: spec/system/candidate_interface/continuous_applications/submitting/candidate_submitting_application_spec.rb [hunk 2]
```diff
@@ -18,7 +20,7 @@
     then_i_see_a_interruption_page_for_personal_statement
 
     when_i_continue_without_editing
-    then_i_should_see_the_review_and_submit_page
+    then_i_should_be_on_the_review_and_submit_page(application_choice: @application_choice)
```
Comment (inulty-dfe) [inline spec/system/candidate_interface/continuous_applications/submitting/candidate_submitting_application_spec.rb:23]: I would suggest we remove `should` from all methods. It's redundant and is not assertive.

Comment (inulty-dfe) [reply spec/system/candidate_interface/continuous_applications/submitting/candidate_submitting_application_spec.rb:23]: Also, we should be using the instance variable and not passing arguments to these methods unless it is somehow necessary

File: spec/system/candidate_interface/continuous_applications/submitting/candidate_tries_to_submit_an_application_for_a_course_with_closed_application_spec.rb [hunk 3]
```diff
@@ -0,0 +1,65 @@
+require 'rails_helper'
+
+RSpec.feature 'Candidate tries to submit an application choice which the course is closed' do
```
Comment (inulty-dfe) [inline spec/system/candidate_interface/continuous_applications/submitting/candidate_tries_to_submit_an_application_for_a_course_with_closed_application_spec.rb:3]:
```suggestion
RSpec.feature 'Candidate tries to submit an application choice when the course is closed' do
```

File: spec/system/candidate_interface/continuous_applications/submitting/candidate_tries_to_submit_an_application_for_course_unavailable_spec.rb [hunk 4]
```diff
@@ -0,0 +1,74 @@
...
+
+RSpec.feature 'Candidate tries to submit an application choice which the course is unavailable' do
+  include SignInHelper
+  include CandidateSubmissionHelper
+  before do
+    given_i_am_signed_in
+    and_i_have_one_application_in_draft
+  end
+
+  after do
```
Comment (inulty-dfe) [inline spec/system/candidate_interface/continuous_applications/submitting/candidate_tries_to_submit_an_application_for_course_unavailable_spec.rb:11]: I'm not a fan of using hooks to deduplicate tests by putting assertions in them.

The hooks are for setup and tear down.

The line that the test fails on will point to this block and not the test that failed. (Although the test line is available)
Maybe use a new method if duplication is a big issue.

File: spec/system/candidate_interface/continuous_applications/submitting/candidate_tries_to_submit_an_application_for_a_course_with_closed_application_spec.rb [hunk 5]
```diff
@@ -0,0 +1,65 @@
...
+  include SignInHelper
+  include CandidateSubmissionHelper
+  before do
+    given_i_am_signed_in
+    and_i_have_one_application_in_draft
+  end
+
+  scenario 'Apply is open but course not open for applications', time: mid_cycle(CycleTimetable.current_year) do
```
Comment (inulty-dfe) [inline spec/system/candidate_interface/continuous_applications/submitting/candidate_tries_to_submit_an_application_for_a_course_with_closed_application_spec.rb:11]: Can we use `CycleTimetableHelper.mid_cycle` to keep all `time: ` options consistent?

**Review by inulty-dfe — COMMENTED** [review]
Great work adding more confidence to this part of the code base.

I'm not sure now is the time to create `CandidateSubmissionHelper`.

I would really like to see an end to `should` in the test suite

File: spec/system/candidate_interface/continuous_applications/submitting/candidate_tries_to_submit_an_application_for_course_unavailable_spec.rb (hunk 4 above)
Comment (tomas-stefano) [reply spec/system/candidate_interface/continuous_applications/submitting/candidate_tries_to_submit_an_application_for_course_unavailable_spec.rb:11]: I can duplicate in all scenarios.

File: spec/support/test_helpers/candidate_submission_helper.rb (hunk 1 above)
Comment (tomas-stefano) [reply spec/support/test_helpers/candidate_submission_helper.rb:1]: I just need to reuse many of the steps and I think adding to CandidateHelper will just make a module where we put everything, hence why I create this module.

File: spec/system/candidate_interface/continuous_applications/submitting/candidate_submitting_application_spec.rb (hunk 2 above)
Comment (tomas-stefano) [reply spec/system/candidate_interface/continuous_applications/submitting/candidate_submitting_application_spec.rb:23]: I kinda of agree with the should, I consciously decided to keep the same pattern as we have many like that in the code:
```
⤷  ag then_i_should spec/ | wc -l
     850
```

Comment (tomas-stefano) [reply spec/system/candidate_interface/continuous_applications/submitting/candidate_submitting_application_spec.rb:23]: About the instance variable, I agree. I will change

File: spec/system/candidate_interface/continuous_applications/submitting/candidate_tries_to_submit_an_application_for_a_course_with_closed_application_spec.rb (hunk 5 above)
Comment (tomas-stefano) [reply spec/system/candidate_interface/continuous_applications/submitting/candidate_tries_to_submit_an_application_for_a_course_with_closed_application_spec.rb]: Done

File: spec/system/candidate_interface/continuous_applications/submitting/candidate_tries_to_submit_an_application_for_course_unavailable_spec.rb (hunk 4 above)
Comment (tomas-stefano) [reply spec/system/candidate_interface/continuous_applications/submitting/candidate_tries_to_submit_an_application_for_course_unavailable_spec.rb]: Done

File: spec/system/candidate_interface/continuous_applications/submitting/candidate_tries_to_submit_an_application_for_a_course_with_closed_application_spec.rb (hunk 3 above)
Comment (tomas-stefano) [reply spec/system/candidate_interface/continuous_applications/submitting/candidate_tries_to_submit_an_application_for_a_course_with_closed_application_spec.rb]: Done

File: spec/system/candidate_interface/continuous_applications/submitting/candidate_submitting_application_spec.rb (hunk 2 above)
Comment (tomas-stefano) [reply spec/system/candidate_interface/continuous_applications/submitting/candidate_submitting_application_spec.rb]: About the instance variable is done.

**Review by inulty-dfe — APPROVED** [review]

**Pull request merged**